src/components/core/Dropdown/Dropdown.tsx

```diff
@@ -0,0 +1,107 @@
+import {
+  ReactNode,
+  RefObject,
+  useCallback,
+  useEffect,
+  useRef,
+  useState,
+} from 'react';
+import styled from 'styled-components';
+import TextButton from '../Button/TextButton';
+import colors from '../colors';
+
+type Props = {
+  mainText: string;
+  children?: ReactNode;
+};
+
+const refContainsEventTarget = (ref: RefObject<HTMLDivElement>, event: any) => {
+  return ref.current && ref.current.contains(event.target);
+};
+
+function Dropdown({ mainText, children }: Props) {
+  const [isDropdownVisible, setIsDropdownVisible] = useState(false);
+
+  const handleClick = useCallback(() => {
+    setIsDropdownVisible(!isDropdownVisible);
+  }, [isDropdownVisible]);
+
+  const dropdownMenuRef = useRef<HTMLDivElement>(null);
+  const dropdownButtonRef = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
+    function handleDocumentClick(event: any) {
+      if (!isDropdownVisible) {
+        return;
+      }
+
+      const clickedOutsideMenu = !refContainsEventTarget(
+        dropdownMenuRef,
+        event
+      );
+      const clickedOnButton = refContainsEventTarget(dropdownButtonRef, event);
+
+      if (clickedOutsideMenu && !clickedOnButton) {
+        setIsDropdownVisible(false);
+      }
+    }
+
+    // we only need to add document listener if dropdown is open
+    if (isDropdownVisible) {
+      document.addEventListener('mousedown', handleDocumentClick);
+    }
+    return () => {
+      document.removeEventListener('mousedown', handleDocumentClick);
+    };
+  }, [dropdownMenuRef, isDropdownVisible]);
```

Comment:
confused about these deps:
1. dropdownMenuRef is a react ref, which means it shouldn't need to be included as a dependency 🤔 
2. if `isDropdownVisible` is part of the dep array, does it mean whenever the state changes we run the effect? ideally we only run this effect once. i think i gotta study this a bit more

Reply:
yeaah we can discuss in person.
I debated the "set listener only once, on component mount" vs this approach

With this approach, the document listener gets set when the dropdown opens, and is removed when it's closed. That way it's not listening for document clicks when we don't care about it. It seems a bit extra compared to setting the listener once and forgetting about it, but it also feels cleaner to me.

That was also part of the reason for not passing in a `isDropdownVisible` argument to handleDocumentClick - we should only need to handleDocumentClick if the dropdown is open/visible, so if it's called only when necessary, `isDropdownVisible` will/should always be true.

But I also like your suggestion to memoize the function and want to make sure I'm adhering to best practices. a further optimization could be to make it so the effect only runs if `isDropdownVisible === true`. We can discuss and i'll continue tinkering